File: src/ros2_bridge/mappers/geometry_msgs/twist.rs
//! Typed mapper for `geometry_msgs/msg/Twist`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn twist_to_bus(msg: ros_env::geometry_msgs::msg::Twist) -> crate::geometry_msgs::msg::v1::Twist {
    crate::geometry_msgs::msg::v1::Twist {
        linear: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.linear)),
        angular: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.angular)),
    }
}

pub(crate) fn twist_to_ros(bus: crate::geometry_msgs::msg::v1::Twist) -> ros_env::geometry_msgs::msg::Twist {
    ros_env::geometry_msgs::msg::Twist {
        linear: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(bus.linear.unwrap_or_default()),
        angular: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(bus.angular.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsTwistMapper;

impl TypedTopicMapper for GeometryMsgsTwistMapper {
    type Ros = ros_env::geometry_msgs::msg::Twist;
    type Bus = crate::geometry_msgs::msg::v1::Twist;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/Twist"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(twist_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(twist_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/twist.py
"""Generated mapper for `geometry_msgs/msg/Twist`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.vector3 import vector3_to_bus, vector3_to_ros

def twist_to_bus(msg):
    from robot_bus.geometry_msgs.msg.v1 import Twist as BusMsg

    bus = BusMsg()
    bus.linear.CopyFrom(vector3_to_bus(msg.linear))
    bus.angular.CopyFrom(vector3_to_bus(msg.angular))
    return bus


def twist_to_ros(bus):
    from geometry_msgs.msg import Twist as RosMsg

    out = RosMsg()
    out.linear = vector3_to_ros(bus.linear)
    out.angular = vector3_to_ros(bus.angular)
    return out


class GeometryMsgsTwistMapper:
    def ros_msg_type(self):
        from geometry_msgs.msg import Twist as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return twist_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.geometry_msgs.msg.v1 import Twist as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return twist_to_ros(bus)
